src/company_gui.cpp

```diff
@@ -1827,267 +1845,217 @@ struct CompanyInfrastructureWindow : Window
-			DrawString(tr.left, tr.right, y,
-				TimerGameEconomy::UsingWallclockUnits() ? STR_COMPANY_INFRASTRUCTURE_VIEW_TOTAL_PERIOD : STR_COMPANY_INFRASTRUCTURE_VIEW_TOTAL_YEAR,
-				TC_FROMSTRING, SA_RIGHT);
+			SetDParam(0, max_cost * 12);
+			max_cost_width = GetStringBoundingBox(TimerGameEconomy::UsingWallclockUnits()
+					? STR_COMPANY_INFRASTRUCTURE_VIEW_TOTAL_PERIOD : STR_COMPANY_INFRASTRUCTURE_VIEW_TOTAL_YEAR).width;
+
+			size.height   += WidgetDimensions::unscaled.vsep_normal * 2;
+		} else {
+			max_cost_width = 0;
 		}
+
+		size.height   = Clamp(size.height, 0, GetLineHeight() * 25);   // max 25 lines otherwise expand widow manually
+		size.width    = max_label_width + WidgetDimensions::scaled.hsep_normal + max_cost_width + WidgetDimensions::scaled.hsep_normal + max_count_width;
+
+		resize.height = GetLineHeight();
 	}
 
 	void DrawWidget(const Rect &r, WidgetID widget) const override
 	{
-		const Company *c = Company::Get((CompanyID)this->window_number);
+		if (widget != WID_CI_LIST) return;
 
-		int y = r.top;
+		int y = r.top + 1;
+		Rect rr = r.Shrink(WidgetDimensions::scaled.hsep_normal);
+		Rect ir = rr.Indent(WidgetDimensions::scaled.hsep_indent, _current_text_dir == TD_RTL);
 
-		Rect ir = r.Indent(WidgetDimensions::scaled.hsep_indent, _current_text_dir == TD_RTL);
-		switch (widget) {
-			case WID_CI_RAIL_DESC:
-				DrawString(r.left, r.right, y, STR_COMPANY_INFRASTRUCTURE_VIEW_RAIL_SECT);
-
-				if (this->railtypes != RAILTYPES_NONE) {
-					/* Draw name of each valid railtype. */
-					for (const auto &rt : _sorted_railtypes) {
-						if (HasBit(this->railtypes, rt)) {
-							DrawString(ir.left, ir.right, y += GetCharacterHeight(FS_NORMAL), GetRailTypeInfo(rt)->strings.name, TC_WHITE);
-						}
-					}
-					DrawString(ir.left, ir.right, y += GetCharacterHeight(FS_NORMAL), STR_COMPANY_INFRASTRUCTURE_VIEW_SIGNALS);
-				} else {
-					/* No valid railtype. */
-					DrawString(ir.left, ir.right, y += GetCharacterHeight(FS_NORMAL), STR_COMPANY_VIEW_INFRASTRUCTURE_NONE);
-				}
+		int label_right = rr.left + this->max_label_width;          // resize rect to handle text direction change
+		int count_left  = ir.right - this->max_count_width;
+		int cost_right  = count_left - WidgetDimensions::scaled.hsep_normal;
+		int cost_left   = cost_right - this->max_cost_width;
 
-				break;
 
-			case WID_CI_RAIL_COUNT: {
-				/* Draw infrastructure count for each valid railtype. */
-				uint32_t rail_total = c->infrastructure.GetRailTotal();
-				for (const auto &rt : _sorted_railtypes) {
-					if (HasBit(this->railtypes, rt)) {
-						this->DrawCountLine(r, y, c->infrastructure.rail[rt], RailMaintenanceCost(rt, c->infrastructure.rail[rt], rail_total));
-					}
-				}
-				if (this->railtypes != RAILTYPES_NONE) {
-					this->DrawCountLine(r, y, c->infrastructure.signal, SignalMaintenanceCost(c->infrastructure.signal));
-				}
-				break;
-			}
+		this->BuildCompanyInfrastructureList();
```

Comment:
Does it need to be updated every redraw? Seems you've called it a "cache" but as it's rebuilt every draw it's not much of one. It might that rebuild could be trigged in OnInvalidateData, perhaps be timer-based. (Stats updating by timer every 1 to 3 seconds is quite common)

Reply:
A timer 1-3 seconds can work.

This approach will be different from the original were the stats were updated on every call to `DrawWidget`.
```
	case WID_CI_RAIL_COUNT: {
		/* Draw infrastructure count for each valid railtype. */
		uint32_t rail_total = c->infrastructure.GetRailTotal();
		for (const auto &rt : _sorted_railtypes) {
			if (HasBit(this->railtypes, rt)) {
				this->DrawCountLine(r, y, c->infrastructure.rail[rt], RailMaintenanceCost(rt, c->infrastructure.rail[rt], rail_total));
			}
		}
		if (this->railtypes != RAILTYPES_NONE) {
			this->DrawCountLine(r, y, c->infrastructure.signal, SignalMaintenanceCost(c->infrastructure.signal));
		}
		break;
	}

	case WID_CI_ROAD_DESC:
	case WID_CI_TRAM_DESC: {
		DrawString(r.left, r.right, y, widget == WID_CI_ROAD_DESC ? STR_COMPANY_INFRASTRUCTURE_VIEW_ROAD_SECT : STR_COMPANY_INFRASTRUCTURE_VIEW_TRAM_SECT);

		/* Draw name of each valid roadtype. */
		for (const auto &rt : _sorted_roadtypes) {
			if (HasBit(this->roadtypes, rt) && RoadTypeIsRoad(rt) == (widget == WID_CI_ROAD_DESC)) {
				DrawString(ir.left, ir.right, y += GetCharacterHeight(FS_NORMAL), GetRoadTypeInfo(rt)->strings.name, TC_WHITE);
			}
		}

		break;
	}
```

Reply:
Yup. This one was just an extra idea, as you have now 'cached' the values, it doesn't actually need to recalculate every draw. But maybe somebody would find a delay awkward.